### ISSUE TITLE
Provide service name for interpolation

### ISSUE THREAD

**Issue by ciaranmcnulty** [compose-spec/compose-spec#263]
**What is the problem you're trying to solve**

Yaml configuration can often have boilerplate, differing only by the service name. This is a small example, but in real life the repetition can be much greater:

```yaml
services:
  web:
    image: "${REGISTRY}/web:${TAG}"
    build:
      cache_from: "${REGISTRY}/web:main"
  worker:
    image: "${REGISTRY}/worker:${TAG}"
    build:
      cache_from: "${REGISTRY}/worker:main"
  [... more similar services ...]]
````

**Describe the solution you'd like**

Similar to [how the project name is exposed](https://github.com/compose-spec/compose-spec/blob/5260c1bc8ee5e93ce88412bb460924e9535d9080/spec.md#name-top-level-element), expose the service name for interpolation. 

Ideally interpolation would happen after Yaml anchors are resolved, to enable this:

```yaml
services:
  web: &template-service
    image: "${REGISTRY}/${COMPOSE_SERVICE_NAME}:${TAG}"
    build:
      cache_from: "${REGISTRY}/${COMPOSE_SERVICE_NAME}:main"
  worker: *template-service
````

**Additional context**

I made a similar request to `buildx bake`'s HCL format https://github.com/docker/buildx/issues/1156

**Comment by EricHripko** [comment]
Hi @ciaranmcnulty; thank you for sharing this feedback 👍

**Comment by ciaranmcnulty** [comment]
What's the best way to progress a proposal? I should rephrase this as a PR to the spec?

**Comment by EricHripko** [comment]
Apologies for the delay in getting back to you!

I would expect that the first step would be to get buy-in from maintainers (this overlaps a lot with the Compose tooling authors) on whether this is a problem that Compose spec should be solving. Paging @ndeloof, @hangyan, @justincormack, @ulyssessouza, @glours for their opinion on this 👀

**Comment by ulyssessouza** [comment]
I understand the usecase, but that would introduce the idea of having a variable that has a different values depending on its context in the file. For now, `compose-spec/compose-go` does its interpolation in the plain file, so in a non contextualised way.

> Note that by this I don't mean that the spec should be driven by the actual implementation. I'm just saying that it would at least increase its complexity.

**Comment by ciaranmcnulty** [comment]
@ulyssessouza ah so interpolation happens before Yaml parsing?

 I guess you'd need to have a var that _wasn't_ interpolated at that stage, and then a separate pass on the parsed structure once the service names are known (and anchors are resolved etc).

Maybe a new-Bash syntax would work? I don't know if that's more or less confusing :)

For best DX this would get interpolated in `docker compose config`

**Comment by ciaranmcnulty** [comment]
I'd also love to discuss other solutions to the same use case (boilerplate with slight variations) - functions like in bake's HCL?
